Sources/SentryCrash/Recording/Tools/SentryHook.c

```diff
@@ -0,0 +1,227 @@
+#include "SentryHook.h"
+#include "fishhook.h"
+#include <dispatch/dispatch.h>
+#include <execinfo.h>
+#include <mach/mach.h>
+#include <pthread.h>
+
+// NOTE on accessing thread-locals across threads:
+// We save the async stacktrace as a thread local when dispatching async calls,
+// but the various crash handlers need to access these thread-locals across threads
+// sometimes, which we do here:
+// While `pthread_t` is an opaque type, the offset of `thread specific data` (tsd)
+// is fixed due to backwards compatibility.
+// See:
+// https://github.com/apple/darwin-libpthread/blob/c60d249cc84dfd6097a7e71c68a36b47cbe076d1/src/types_internal.h#L409-L432
+
+#if __LP64__
+#    define TSD_OFFSET 224
+#else
+#    define TSD_OFFSET 176
+#endif
+
+static pthread_key_t async_caller_key = 0;
+
+sentrycrash_async_backtrace_t *
+sentrycrash_get_async_caller_for_thread(SentryCrashThread thread)
+{
+    const pthread_t pthread = pthread_from_mach_thread_np((thread_t)thread);
+    void **tsd_slots = (void *)((uint8_t *)pthread + TSD_OFFSET);
+
+    return (sentrycrash_async_backtrace_t *)__atomic_load_n(
+        &tsd_slots[async_caller_key], __ATOMIC_SEQ_CST);
+}
+
+void
+sentrycrash__async_backtrace_incref(sentrycrash_async_backtrace_t *bt)
+{
+    if (!bt) {
+        return;
+    }
+    __atomic_fetch_add(&bt->refcount, 1, __ATOMIC_SEQ_CST);
```

Comment:
i suspect SEQ_CST is overkill, but this is hard so probably fine to leave it until this is actually a perf bottleneck

Reply:
We use basically the same all over the native SDK. Getting the memory orderings right is quite difficult, so IMO there is nothing wrong with being too conservative.

BTW, there is a nice series on these patterns on LWN which I’m reading these days: https://lwn.net/Articles/844224/